File: data_structure/Balanced_binary_tree/recursive.py
class Solution:
    def isBalanced(self, root):  #recursive
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        def depth(node):
            if not node:
                return 0 # leaves has no length
            left=depth(node.left)
            right=depth(node.right)
            if abs(left-right) > 1:
                raise Exception
            return max(left,right)+1
        try:
            return abs(depth(root.left)-depth(root.right))<=1
        except:
            return False

File: data_structure/Balanced_binary_tree/recursive.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    struct Error : public std::exception {};   //using standard exception type
    int depth(TreeNode* node){
        if(!node){
            return 0;
        }
        int left = depth(node->left);
        int right = depth(node->right);
        if(abs(left-right)>1){
            throw Error();
        }
        return max(left,right)+1;
    }
    bool isBalanced(TreeNode* root) {
        if(!root){
            return true;
        }
        try
        {
            return abs(depth(root->left)-depth(root->right))<=1;
        }catch(Error e)
        {
            return false;
        }
        
    }
};